scipy/spatial/distance.py

```diff
@@ -679,7 +679,7 @@ def cosine(u, v, w=None):
     # cosine distance is also referred to as 'uncentered correlation',
     #   or 'reflective correlation'
     # clamp the result to 0-2
-    return max(0, min(correlation(u, v, w=w, centered=False), 2.0))
+    return max(0.0, min(correlation(u, v, w=w, centered=False), 2.0))
```

Comment:
`correlation` already seems to return the absolute value so I guess we might be able to completely remove the call to `max` here.

https://github.com/scipy/scipy/blob/ae9dc13161cb227d7651c58dc47971193510221d/scipy/spatial/distance.py#L636

Reply:
Perhaps you could replace the `np.abs` call in `correlation` with `np.clip` and remove the additional fixup from `cosine`.

Reply:
> Perhaps you could replace the np.abs call in correlation with np.clip and remove the additional fixup from cosine.

Do you want to try doing this @liuzhenqi77 ?

Reply:
Sorry for the delay! It is fixed now using `np.clip`, but should I change it to builtin max/min in light of recent updates by @lgeiger ?

Reply:
Sure, you're likely to see a small speed improvement from using builtins for scalar values. It would also be good to add a small regression test that checks the output type of correlation and cosine.